web/packages/teleport/src/Main/Main.tsx

```diff
@@ -169,13 +171,32 @@ export function Main(props: MainProps) {
   );
 }
 
-function renderRoutes(features: TeleportFeature[]) {
+function renderRoutes(features: TeleportFeature[], ctx: TeleportContext) {
   const routes = [];
 
   for (const [index, feature] of features.entries()) {
     if (feature.route) {
       const { path, title, exact, component: Component } = feature.route;
 
+      // add regular feature routes
+      routes.push(
+        <Route title={title} key={index} path={path} exact={exact}>
+          <CatchError>
+            <Suspense fallback={null}>
+              <Component />
+            </Suspense>
+          </CatchError>
+        </Route>
+      );
+    }
+
+    // add the route of the 'locked' variants of the features
+    if (feature.isLocked !== undefined && feature.isLocked(ctx)) {
+      if (!feature.lockedRoute) {
+        continue;
+      }
+
+      const { path, title, exact, component: Component } = feature.lockedRoute;
```

Comment:
hm, the children should probably have the `isLocked` check too. so something like 

```ts
if (feature.isLocked && !feature.parent) {
  // register route
} // ignore children of locked parent
```

b/c if we allow the routes to be registered, a user can manipulate the route (or go to a possibly a saved URL) and it will still render the screen.

also do we need to pass the context? i think it's okay to call `useTeleport` instead of passing it down.

Reply:
or i wonder if we can keep an array of locked `parents`, and check if the route has a parent, check if it's an instance of the locked parent? something like this 
https://github.com/gravitational/teleport/blob/0e8ae872f5140626febd597cc22dfa0873cbd865/web/packages/teleport/src/Navigation/NavigationDropdown.tsx#L143

Reply:
> also do we need to pass the context? i think it's okay to call useTeleport instead of passing it down.

We get an eslint warning if we call `useTeleport` in `renderRoutes`:

`React Hook "useTeleport" is called in function "renderRoutes" that is neither a React function component nor a custom React Hook function. React component names must start with an uppercase letter. React Hook names must start with the word "use".eslint[react](https://reactjs.org/docs/hooks-rules.html)`

Reply:
> or i wonder if we can keep an array of locked `parents`, and check if the route has a parent, check if it's an instance of the locked parent? something like this (removing the need to define a isLock check for childrens)
> 
> https://github.com/gravitational/teleport/blob/0e8ae872f5140626febd597cc22dfa0873cbd865/web/packages/teleport/src/Navigation/NavigationDropdown.tsx#L143

We could do this:

```typescript
function getLockedParents(
  features: TeleportFeature[],
  ctx: TeleportContext
): string[] {
  const lockedParents = new Set<string>();
  features.forEach(feature => {
    if (feature.parent) {
      const parent = new feature.parent();
      if (parent.isLocked(ctx)) {
        lockedParents.add(parent.constructor.name);
      }
    }
  });

  return Array.from(lockedParents);
}

function renderRoutes(features: TeleportFeature[], ctx: TeleportContext) {
  const routes = [];
  const lockedParents = getLockedParents(features, ctx);

  for (const [index, feature] of features.entries()) {
    const isParentLocked = lockedParents.find(
      parent => feature.parent && parent === feature.parent.name
    );
    if (isParentLocked) {
      continue;
    }
  // ...
}
```

Not pretty, but we have to know somehow that the feature's parent is locked if we want to remove that route.

Reply:
i think we can simplify this, can you try this out:

os: https://github.com/gravitational/teleport/commit/dfdf539b8073335e1b23e2c9b943ce1f13a8262a
e: https://github.com/gravitational/teleport.e/commit/2d1958e5b23056e66c61c36c8e4864572fc1ef25

Reply:
I like this, thank you.

Added in https://github.com/gravitational/teleport/pull/25588/commits/3e5d91ecc93b73090f823d63e1fc2ac75692e26b and updated on the `e` PR too.

Reply:
I have some concerns about `isLockedAndUpdatedRouteAndNavigationItem`:
- the name tells that the method only returns a value, but that's not true; it also modifies the feature,
- it is called in `useMemo`, and it is not a good practice to call side effects in it https://stackoverflow.com/a/62507767
> You may rely on useMemo as a performance optimization, not as a semantic guarantee. In the future, React may choose to “forget” some previously memoized values and recalculate them on next render, e.g. to free memory for offscreen components. Write your code so that it still works without useMemo — and then add it to optimize performance. 

It seems to me that we should modify the features in a different place (for example, in `ctx.init()`) and in `useMemo` only check `isLocked` value. 

But TBH, I like the first version of this PR, where `TeleportFeature` was a simple class that didn't perform any logic. 

Also, didn't we lose this change after the refactor? https://github.com/gravitational/teleport/blob/4f568e7318f4767fe0c7b884da1d370bf51fc560/web/packages/teleport/src/Navigation/NavigationItem.tsx#L203

Reply:
Right, thank you.

We could make `isLockedAndUpdatedRouteAndNavigationItem` become two separate methods: `isLocked` and `setRouteAndNavitaionItemAsLock` or something.

Then, on `Main.tsx`, we filter adjust the routes and nav items based on the context and use it in the useMemo:

```typescript
 // use context to determine which features should be removed
  // due to being unavailable on the cluster
  const correctedRouteFeatures = props.features.map(feature => {
    if (
      feature.isLocked && // check if isLocked is defined since it is optional
      feature.isLocked(ctx.lockedFeatures)
    ) {
      feature.setRouteAndNavitaionItemAsLock();
    }
    return feature;
  });

  const features = useMemo(() => {
    const lockedParentFeatures: TeleportFeature[] =
      correctedRouteFeatures.filter(feature => feature.hasAccess);

    // remove features which the parent feature is locked
    return correctedRouteFeatures.filter(feature => {
      if (!feature.hasAccess(featureFlags)) {
        return false;
      }
      if (!feature.parent) {
        return true;
      }
      // This feature is a child of a parent feature.
      // Check if the parent is locked, which means
      // the children should also be locked from access.
      return !lockedParentFeatures.some(
        lockedParent => lockedParent instanceof feature.parent
      );
    });
  }, [featureFlags, correctedRouteFeatures]);
```

Reply:
The question is where to create `correctedRouteFeatures`. You can't just place it in the rendering function, because it will make `useMemo`  run over and over. We could do it in `ctx.init()` but since we mutate the classes React may not see the changes and `useMemo` won't re-run. 
I feel we could avoid these questions by not mutating `features` and instead just calculating the needed value while rendering, like this: 
`feature.isLocked() ? feature.lockedRoute : feature.route`.

(it is more or less something you had before)

Reply:
sounds good to me 👍